src/controller/python/chip/ble/LinuxImpl.cpp

```diff
@@ -11,41 +12,126 @@ using namespace chip::DeviceLayer::Internal;
 {
-    return static_cast<chip::DeviceLayer::Internal::AdapterIterator *>(adapter)->Next();
+    return static_cast<AdapterIterator *>(adapterIterator)->Next();
 }
 
-extern "C" unsigned pychip_ble_adapter_list_get_index(void * adapter)
+extern "C" unsigned pychip_ble_adapter_list_get_index(void * adapterIterator)
 {
     /// NOTE: returning unsigned because python native has no sized values
-    return static_cast<chip::DeviceLayer::Internal::AdapterIterator *>(adapter)->GetIndex();
+    return static_cast<AdapterIterator *>(adapterIterator)->GetIndex();
 }
 
-extern "C" const char * pychip_ble_adapter_list_get_address(void * adapter)
+extern "C" const char * pychip_ble_adapter_list_get_address(void * adapterIterator)
 {
-    return static_cast<chip::DeviceLayer::Internal::AdapterIterator *>(adapter)->GetAddress();
+    return static_cast<AdapterIterator *>(adapterIterator)->GetAddress();
 }
 
-extern "C" const char * pychip_ble_adapter_list_get_alias(void * adapter)
+extern "C" const char * pychip_ble_adapter_list_get_alias(void * adapterIterator)
 {
-    return static_cast<chip::DeviceLayer::Internal::AdapterIterator *>(adapter)->GetAlias();
+    return static_cast<AdapterIterator *>(adapterIterator)->GetAlias();
 }
 
-extern "C" const char * pychip_ble_adapter_list_get_name(void * adapter)
+extern "C" const char * pychip_ble_adapter_list_get_name(void * adapterIterator)
 {
-    return static_cast<chip::DeviceLayer::Internal::AdapterIterator *>(adapter)->GetName();
+    return static_cast<AdapterIterator *>(adapterIterator)->GetName();
 }
 
-extern "C" bool pychip_ble_adapter_list_is_powered(void * adapter)
+extern "C" bool pychip_ble_adapter_list_is_powered(void * adapterIterator)
 {
-    return static_cast<chip::DeviceLayer::Internal::AdapterIterator *>(adapter)->IsPowered();
+    return static_cast<AdapterIterator *>(adapterIterator)->IsPowered();
+}
+
+extern "C" void * pychip_ble_adapter_list_get_raw_adapter(void * adapterIterator)
+{
+    return static_cast<AdapterIterator *>(adapterIterator)->GetAdapter();
+}
+
+/////////// CHIP Device scanner implementation //////////
+
+namespace {
+
+// To avoid pythoon compatibility issues on inc/dec references,
+// code assumes an abstract type and leaves it up to python to keep track of
+// reference counts
+struct PyObject;
+
+class ScannerDelegateImpl : public ChipDeviceScannerDelegate
+{
+public:
+    using DeviceScannedCallback = void (*)(PyObject * context, const char * address, uint16_t discriminator, uint16_t vendorId,
+                                           uint16_t productId);
+    using ScanCompleteCallback  = void (*)(PyObject * context);
+
+    ScannerDelegateImpl(PyObject * context, DeviceScannedCallback scanCallback, ScanCompleteCallback completeCallback) :
+        mContext(context), mScanCallback(scanCallback), mCompleteCallback(completeCallback)
+    {}
+
+    ~ScannerDelegateImpl()
+    {
+        mContext          = nullptr;
```

Comment:
Debugging invalid memory references. Turned out to be python c_void_p behavind oddly (it casts to 32-bit if it can, but when passing back it stays 32-bit int and does a sign extension or invalid high order bits when sent as a function argument ... very frustrating)